src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,249 @@
+    | Some x -> is_rec_typ typ_names x
+    | None -> false
+  in
+  in_type_kind || in_manifest
+
+and is_rec_typ typ_names = function
   | { ptyp_desc = Ptyp_constr ({ txt = x; _ }, _); _ } ->
-      longident_to_str x = typ_name
-  | { ptyp_desc = Ptyp_tuple xs; _ } -> List.exists (is_rec_typ typ_name) xs
+     List.exists (fun typ_name -> longident_to_str x = typ_name) typ_names
+  | { ptyp_desc = Ptyp_tuple xs; _ } -> List.exists (is_rec_typ typ_names) xs
+  | { ptyp_desc = Ptyp_variant (rws, _, _); _ } ->
+     List.exists (fun rw ->
+         match rw.prf_desc with
+         | Rtag (lab, _, cts) ->
+            List.exists (fun typ_name -> lab.txt = typ_name) typ_names ||
+              List.exists (is_rec_typ typ_names) cts
+         | Rinherit ct -> is_rec_typ typ_names ct) rws
   | _ -> false
 
-let name s =
-  let prefix = "gen" in
-  match s with "t" -> prefix | s -> prefix ^ "_" ^ s
+and is_rec_constr_decl typ_names cd =
+  match cd.pcd_args with
+  | Pcstr_tuple cts -> List.exists (is_rec_typ typ_names) cts
+  | _ -> false
 
-let pat ~loc s =
-  let (module A) = Ast_builder.make loc in
-  let s = name s in
-  A.pvar s
+(** [env] contains the list of recursive types during the derivation
+
+    i.e:
+    - contains one element maximum if its a single type declaration *)
+type env = string list ref
+
+(** [curr_types] saves both the current type and all mutual
+    recursive types *)
+let curr_types : (string * string list) option ref = ref None
+
+let set_curr_types x xs = curr_types := Some (x, xs)
+
+let clean_curr_types () = curr_types := None
 
-let gen ~loc ?(env = TypeGen.empty) lg =
+let curr_type () = Option.get !curr_types |> fst
+
+let curr_types () = Option.get !curr_types |> snd
+
+let env : env = ref []
+
+let clean_env () = env := []
+
+let is_rec x = List.mem x !env
+
+let add_env typ_names typ_name ty =
+  if is_rec_type_decl typ_names ty then
+    env := typ_name :: !env
+
+
+(** {2. Generator constructors} *)
+
+(** [gen lg args] creates a generator using [lg].
+
+    The longident can either be a:
+    - Lident s: We transform to gen_s (or gen if s = "t")
+    - Ldot (lg, s): We transform to qualified generator (e.g. B.gen)
+*)
+let gen ~loc lg args =
   let (module A) = Ast_builder.make loc in
   match lg with
   | Lident s ->
-      Option.value ~default:(name s |> A.evar) @@ TypeGen.find_opt s env
-  | Ldot (lg, s) -> A.(pexp_ident (Located.mk @@ Ldot (lg, name s)))
+     if is_rec s then
+       name ~sized:true s |> A.evar |>
+         Args.apply_args ~loc args |>
+         Args.apply_args ~loc [ [%expr (n / 2)] ]
+     else
+       name s |> A.evar |> Args.apply_args ~loc args
+  | Ldot (lg, s) ->
+     A.(pexp_ident (Located.mk @@ Ldot (lg, name s))) |>
+       Args.apply_args ~loc args
   | Lapply (_, _) -> raise (Invalid_argument "gen received an Lapply")
 
-let frequency ~loc l = [%expr QCheck.Gen.frequency [%e l]]
+(** [tree nodes leaves] creates a tree like generator *)
```

Comment:
I was first confused by `tree`, thinking "why are those type cases handled as a special case?"
After having read further and returning, I can now follow this conceptually (I like trees too!).
However, cut-to-the-bone: this is just a helper function.
It could be inlined in `sized` (there's morally max 2 lines of functionality)

Reply:
Done in 71b284f3fb7e5105d3ce2e84733126a77c16b13c.